Lab3/Lab3/Point.cs

```diff
@@ -1,44 +1,95 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Text.RegularExpressions;
 
 namespace Lab3
 {
     public class Point
     {
         private double x, y;
+        private double initialX, initialY;
 
-        #region Undo_functionality
-        //TODO
-        // tinem un istoric al valorilor pe care le-a avut punctul
-        Stack<Point> history = new Stack<Point>();
+        #region History
+        
+        List<Point> history = new List<Point>();
+        int currentHistoryIndex = 0;
 
-        public void StepBack()
+        public void returnToInitialValues()
         {
+            this.x = initialX;
+            this.y = initialY;
+        }
 
+        public void StepBack(int steps) // how many steps back do we take?
+        {
+            if (currentHistoryIndex < steps)
+            {
+                currentHistoryIndex = 0;
+                return;
+            }
+            else
+            {
+                currentHistoryIndex = currentHistoryIndex - steps;
+
+                this.x = history[currentHistoryIndex].x;
+                this.y = history[currentHistoryIndex].y;
+            }
         }
+
+        public void StepAhead(int steps) // redo
+        {
+            if (currentHistoryIndex + steps >= history.Count)
+            {
+                currentHistoryIndex = history.Count - 1;
+
+                this.x = history[currentHistoryIndex].x;
+                this.y = history[currentHistoryIndex].y;
+            }
+            else
+            {
+                currentHistoryIndex += steps;
+
+                this.x = history[currentHistoryIndex].x;
+                this.y = history[currentHistoryIndex].y;
+            }
+        }
+
         #endregion
         #region c-tors
         public Point(): this(0.0, 0.0)
         {
-
+            
         }
         public Point(double x, double y)
         {
             this.x = x;
             this.y = y;
+
+            initialX = x;
+            initialY = y;
         }
 
         /// <summary>
         /// initializeaza un Point pe baza unui string de forma "(3.0;4.0)"
         /// </summary>
         /// <param name="str"></param>
-        public Point(string str)
+        public Point(string p)
         {
-            // TODO 
-            str = str.Trim();
-
-            // creati un Regex care verifica daca stringrul are forma potrivita
-            
+            // we get rid of the usless elements ( '(' , ')', spaces) and split the string in two => left part of ',' and right part of ','
```

Comment:
first you have to check if your string is  a valid Point, only afterwards you can process it.

Reply:
i used try-catch to deal with invalid Points, I don't know how I could make sure that the input is valid, the possibilities are endless for bad inputs